pages/src/TodoList.js

```diff
@@ -1,124 +1,135 @@
+import React from 'react'
 import { createSelector } from 'reselect'
+import { injectState, provideState, mergeIntoState } from "freactal";
+import { Map } from 'immutable'
 
-export default class TodoList extends Component {
-  static propTypes = {
-    onCreateTodo: PropTypes.func.isRequired,
-    onEditTodo: PropTypes.func.isRequired,
-    todos:ImmutablePropTypes.mapOf(
-      PropTypes.shape({
-        completed: PropTypes.bool.isRequired,
-        label: PropTypes.string.isRequired,
-        id: PropTypes.string.isRequired
-      }).isRequired,
-      PropTypes.string.isRequired
-    )
-  }
+const extractEventValue = fn => (event, ...args) => fn(event.target.value, ...args)
+const extractEventTarget = fn => (event, ...args) => fn(event.target, ...args)
+const eventPreventDefault = fn => (event, ...args) => {
+  event.preventDefault()
+  return fn(event, ...args)
+}
 
-  state = {
+const withState = provideState({
+  initialState: () => ({
     display: 'all',
     newTodoLabel: ''
-  }
-
-  _createTodo = event => {
-    event.preventDefault()
+  }),
+  effects: {
+    initialize: (effects) => fetch('//localhost:20431/todos')
+      .then(res => res.json()
+        .then(todos => mergeIntoState({
+          todos: new Map().withMutations(map => {
+            todos.forEach(todo => {
+              map.set(todo.id, todo)
+            })
+          })
+        })
+        )),
+    sendToServer: (effects, todo) => fetch(`http://localhost:20431/todos/${todo.id}`, {
+      body: JSON.stringify(todo),
+      headers: {
+        'content-type': 'application/json'
+      },
+      method: 'PUT'
+    }),
+    onCreateTodo: (effects) => state => {
 
-    const newTodoLabel = this.state.newTodoLabel.trim()
-    if (newTodoLabel === '') {
-      return
-    }
-
-    this.setState({ newTodoLabel: '' })
+      if (state.newTodoLabel === '') { return }
 
-    // all items of collection should have a unique identifier
-    const id = Math.random().toString(36).slice(2)
-    const todo = {
-      completed: false,
-      id,
-      label: newTodoLabel
+      const id = Math.random().toString(36).slice(2)
+      const todo = {
+        completed: false,
+        id,
+        label: state.newTodoLabel
+      }
+      effects.sendToServer(todo)
+      effects.updateTodo(todo)
+      effects.freeInput()
+    },
+    freeInput: () => state => ({
+      ...state,
+      newTodoLabel: ''
+    }),
+    setNewTodoLabelChange: (_, value) => state =>
+      ({
+        ...state,
+        newTodoLabel: value
+      }),
+    onDisplayChange: (_, value) => state =>
+      ({
+        ...state,
+        display: value
+      }),
+    onTodoCompletionChange: (effects, target) => state => {
+      const { id } = target.dataset
+      effects.updateTodo({
+        ...state.todos.get(id),
+        completed: target.checked
+      })
+      effects.sendToServer({
+        ...state.todos.get(id),
+        completed: target.checked
+      })
```

Comment:
Why not creating a `todo` object once instead of twice? ;)

Reply:
True, I did it on a rush

Reply:
No problem, no need to rush, take all the time you need :)